### PR TITLE
support jwt verify without kid

### PR DESCRIPTION
**Pull request by junnplus**
Signed-off-by: Ye Sijun <junnplus@gmail.com>

> The "kid" (key ID) header parameter is a hint indicating which specific key owned by the signer should be used to validate the signature. This allows signers to explicitly signal a change of key to recipients. Omitting this parameter is equivalent to setting it to an empty string. The interpretation of the contents of the "kid" parameter is unspecified. This header parameter is OPTIONAL.

kid is an optional, so we should be compatible with JWT without kid.

### CODE REVIEW

**Comment by blckngm** [comment]
I am a bit worried about the cost of trying every key in the key set when the set is larger.

Are there real world examples where JWK set is used but JWTs don't have `kid`?

**Comment by junnplus** [comment]
> I am a bit worried about the cost of trying every key in the key set when the set is larger. 

It's up to the party that signed the token to decide, if considering too many keys, the issuer can sign the token with kid.

**Comment by junnplus** [comment]
> Are there real world examples where JWK set is used but JWTs don't have kid?

We have [introduced](https://github.com/datafuselabs/databend/blob/main/query/Cargo.toml#L83) jwtk, but signed JWT has no kid.

**Comment by blckngm** [comment]
> > Are there real world examples where JWK set is used but JWTs don't have kid?
> 
> We have [introduced](https://github.com/datafuselabs/databend/blob/main/query/Cargo.toml#L83) jwtk, but signed JWT has no kid.

So how many keys are in the JWK set? Can we use some heuristics here, say only try every key if there are <= 2 keys.

**Comment by junnplus** [comment]
The current design is to ensure security through the rotation of public keys, maybe two or three.

**Comment by blckngm** [comment]
Or make this opt-in, by adding a method on [`JwkSetVerifier`](https://docs.rs/jwtk/latest/jwtk/jwk/struct.JwkSetVerifier.html#) and [`RemoteJwksVerifier`](https://docs.rs/jwtk/latest/jwtk/jwk/struct.RemoteJwksVerifier.html#) to enable this, but otherwise keep the current behavior (of requiring `kid`).

**Comment by blckngm** [comment]
I'm willing to accept this with either a limit on the number of keys (maybe <= 4) or as an opt-in feature.

File: src/lib.rs [hunk 1]
```diff
@@ -41,6 +42,16 @@ pub struct Header {
     pub extra: Map<String, Value>,
 }
 
+impl TryFrom<&[u8]> for Header {
```
Comment (blckngm) [inline src/lib.rs]: I don't want to expose this `impl TryFrom`. Please make this a private function/method.

**Review by blckngm — CHANGES_REQUESTED** [review]

**Comment by blckngm** [comment]
> I'm willing to accept this with either a limit on the number of keys (maybe <= 4) or as an opt-in feature.

Oh and whichever approach you take, make sure to update the docs to reflect the changes.

File: src/lib.rs (hunk 1 above)
Comment (junnplus) [reply src/lib.rs]: Why? Are there any concerns?

Comment (blckngm) [reply src/lib.rs]: It's a bit weird to have `TryFrom<&[u8]>` when we are actually expecting a text format (JSON). And it's not implemented for `Claims` and `HeadersAndClaims`, making the API surface a bit inconsistent.

**Comment by blckngm** [comment]
Please also rebase so the CI will run. I forgot to include `on: pull_request`.

File: src/lib.rs (hunk 1 above)
Comment (junnplus) [reply src/lib.rs]: > It's a bit weird to have TryFrom<&[u8]> when we are actually expecting a text format (JSON).

How about `TryFrom<&str> `.

> And it's not implemented for `Claims` and `HeadersAndClaims`, making the API surface a bit inconsistent.

We can add the same to `Claims` and `HeadersAndClaims`.

**Comment by junnplus** [comment]
I closed it, I think we'll maintain a fork ourselves. Thanks for your help.

**Comment by blckngm** [comment]
OK. You can also use just `verify` and JWK and handle the JWKS part in your application.

**Pull request closed without merging**